test/jpypetest/test_fault.py

```diff
@@ -0,0 +1,1249 @@
+        _jpype.fault("PyJPNumberLong_repr")
+        with self.assertRaisesRegex(SystemError, "fault"):
+            repr(ji)
+        _jpype.fault("PyJPModule_getContext")
+        with self.assertRaisesRegex(SystemError, "fault"):
+            repr(ji)
+        repr(ji)
+
+    @common.requireInstrumentation
+    def testJPNumberLong_compare(self):
+        ji = JInt(1)
+        _jpype.fault("PyJPNumberLong_compare")
+        with self.assertRaisesRegex(SystemError, "fault"):
+            ji == 1
+        _jpype.fault("PyJPModule_getContext")
+        with self.assertRaisesRegex(SystemError, "fault"):
+            ji == 1
+        ji == 1
+
+    @common.requireInstrumentation
+    def testJPNumberLong_hash(self):
+        ji = JInt(1)
+        _jpype.fault("PyJPNumberLong_hash")
+        with self.assertRaises(SystemError):
+            hash(ji)
+        _jpype.fault("PyJPModule_getContext")
+        with self.assertRaises(SystemError):
+            hash(ji)
+        hash(ji)
+
+    @common.requireInstrumentation
+    def testJPObject_new(self):
+        _jpype.fault("PyJPObject_new")
+        with self.assertRaisesRegex(SystemError, "fault"):
+            JString("a")
+        _jpype.fault("PyJPModule_getContext")
+        with self.assertRaisesRegex(SystemError, "fault"):
+            JString("a")
+        with self.assertRaises(TypeError):
+            _jpype._JObject()
+        JString("a")
+
+    @common.requireInstrumentation
+    def testJPObject_hash(self):
+        jo = JClass("java.lang.Object")()
+        _jpype.fault("PyJPObject_hash")
+        with self.assertRaises(SystemError):
+            hash(jo)
+        _jpype.fault("PyJPModule_getContext")
+        with self.assertRaises(SystemError):
+            hash(jo)
+        hash(jo)
+
+    @common.requireInstrumentation
+    def testJPProxy_new(self):
+        _jpype.fault("PyJPProxy_new")
+        with self.assertRaisesRegex(SystemError, "fault"):
+            JProxy("java.io.Serializable", dict={})
+        with self.assertRaises(TypeError):
+            _jpype._JProxy(None, None)
+        with self.assertRaises(TypeError):
+            _jpype._JProxy(None, [])
+        with self.assertRaises(TypeError):
+            _jpype._JProxy(None, [type])
+        _jpype.fault("JPProxy::JPProxy")
+        with self.assertRaises(SystemError):
+            _jpype._JProxy(None, [JClass("java.io.Serializable")])
+        _jpype._JProxy(None, [JClass("java.io.Serializable")])
+
+# FIXME this needs special treatment
+#    def testJPProxy_str(self):
+#        # Java has a hidden requirement that toString be available
+#        @JImplements("java.util.function.DoubleUnaryOperator")
+#        class f(object):
+#            @JOverride
+#            def applyAsDouble(self, d):
+#                return d
+#        jo = JObject(f(), "java.util.function.DoubleUnaryOperator")
+#        raise RuntimeError(jo.toString())
+
+    @common.requireInstrumentation
+    def testJPProxy_dealloc(self):
```

Comment:
shouldn't this trigger an exception? What am I missing here

Reply:
Assuming you are referring to the deallocation.  Python deallocation routines are special.  Because they trigger underneath everything else and with their own timing.  Therefore, dealloc must eat any exceptions that get triggered or it will result in a Python SystemError thrown at a random time.  Most of the time will be at the end of scope (and thus outside of the test case).   

To handle this properly during testing you will see me using ``def f(): do something;  f()``  which will trigger the action and force a clean up so that we can detect if the call triggers a SystemError inappropriately.  That is also the reason for all the ``assertRaisesRegex`` calls as I need to ensure we get the correct trigger.

Therefore the correct behavior is to trigger a fault and make sure that we DO NOT get a fault because if we do then we will trigger bad behavior in Python (interrupt GC or fail to free memory).

Reply:
Thanks, this makes total sense now. Sometimes tests also deserve documentation (sometimes not).